File: android/src/main/java/expo/modules/localllm/ExpoLocalLlmModule.kt
package expo.modules.localllm

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ExpoLocalLlmModule : Module() {
  private val scope = CoroutineScope(Dispatchers.Main)

  override fun definition() = ModuleDefinition {
    Name("ExpoLocalLlm")

    Events("downloadProgress", "availabilityChange")

    Function("getAvailability") {
      LLMSession.checkAvailability(appContext.reactContext!!).value
    }

    AsyncFunction("downloadModel") Coroutine { ->
      val context = appContext.reactContext!!
      // Immediately signal downloading state so UI can disable the download button
      sendEvent("availabilityChange", mapOf("availability" to ModelAvailability.downloading.value))
      LLMSession.downloadModel(context) { progress ->
        sendEvent("downloadProgress", mapOf("progress" to progress))
      }
      // Re-check availability after download completes
      val newAvailability = LLMSession.checkAvailability(context)
      sendEvent("availabilityChange", mapOf("availability" to newAvailability.value))
    }

    Class(LLMSession::class) {
      Constructor { config: SessionConfig ->
        if (!config.tools.isNullOrEmpty()) {
          throw ToolNotSupportedException()
        }
        LLMSession(appContext.reactContext!!, config)
      }

      AsyncFunction("respond") Coroutine { session: LLMSession, prompt: String ->
        session.respond(prompt)
      }

      AsyncFunction("streamResponse") { session: LLMSession, prompt: String ->
        session.startStream(prompt)
      }

      AsyncFunction("cancelStream") { session: LLMSession ->
        session.cancelStream()
      }

      // Tool calling stubs — not yet supported on Android
      Function("registerTool") { _: LLMSession, _: ToolConfig ->
        throw ToolNotSupportedException()
      }

      Function("unregisterTool") { _: LLMSession, _: String ->
        throw ToolNotSupportedException()
      }

      Function("resolveToolCall") { _: LLMSession, _: String, _: String ->
        throw ToolNotSupportedException()
      }

      Function("rejectToolCall") { _: LLMSession, _: String, _: String ->
        throw ToolNotSupportedException()
      }

      Events("token", "streamComplete", "streamError", "toolCall")
    }

    OnActivityDestroys {
      // Cleanup handled by SharedObject.deallocate() on each LLMSession
    }
  }
}

File: ios/ExpoLocalLlmModule.swift
import ExpoModulesCore

public class ExpoLocalLlmModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoLocalLlm")

    Events("downloadProgress", "availabilityChange")

    Function("getAvailability") { () -> String in
      if #available(iOS 26, *) {
        return LLMSession.checkAvailability().rawValue
      }
      return ModelAvailability.notEligible.rawValue
    }

    AsyncFunction("downloadModel") {
      // No-op on iOS — model is built into the OS
    }

    Class(LLMSession.self) {
      Constructor { (config: SessionConfig) -> LLMSession in
        let session = LLMSession()
        try session.setup(config: config)
        return session
      }

      AsyncFunction("respond") { (session: LLMSession, prompt: String) -> String in
        return try await session.respond(to: prompt)
      }

      AsyncFunction("streamResponse") { (session: LLMSession, prompt: String) in
        try session.startStream(prompt: prompt)
      }

      AsyncFunction("cancelStream") { (session: LLMSession) in
        session.cancelStream()
      }

      // Tool calling support
      Function("registerTool") { (session: LLMSession, config: ToolConfig) in
        session.registerTool(config: config)
      }

      Function("unregisterTool") { (session: LLMSession, name: String) in
        session.unregisterTool(name: name)
      }

      Function("resolveToolCall") { (session: LLMSession, callId: String, result: String) in
        try session.resolveToolCall(callId: callId, result: result)
      }

      Function("rejectToolCall") { (session: LLMSession, callId: String, error: String) in
        try session.rejectToolCall(callId: callId, error: error)
      }
    }

    OnAppEntersForeground {
      // Re-check availability when user returns (may have toggled Apple Intelligence in Settings)
      if #available(iOS 26, *) {
        let current = LLMSession.checkAvailability().rawValue
        self.sendEvent("availabilityChange", [
          "availability": current
        ])
      }
    }

    OnAppEntersBackground {
      // Cancel any active streams when app backgrounds
    }
  }
}
